subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+		if (ret) {
+			LOG_DBG("Get configuration for module %s failed, ret %d", handle->name,
+				ret);
+			return ret;
+		}
+	} else {
+		LOG_DBG("Get configuration for module %s has no get configuration function",
+			handle->name);
+	}
+
+	return 0;
+};
+
+/**
+ * @brief Function to connect two modules together.
+ *
+ */
+int audio_module_connect(struct audio_module_handle *handle_from,
+			 struct audio_module_handle *handle_to)
+{
+	struct audio_module_handle *handle;
+
+	if (handle_from == NULL || handle_to == NULL) {
+		LOG_DBG("Invalid parameter for the connection function");
+		return -EINVAL;
+	}
+
+	if (handle_from->description->type == AMOD_TYPE_UNDEFINED ||
+	    handle_from->description->type == AMOD_TYPE_OUTPUT ||
+	    handle_to->description->type == AMOD_TYPE_UNDEFINED ||
+	    handle_to->description->type == AMOD_TYPE_INPUT) {
+		LOG_DBG("Connections between these modules, %s to %s, is not supported",
+			handle_from->name, handle_to->name);
+		return -ENOTSUP;
+	}
+
+	if (handle_from->state == AMOD_STATE_UNDEFINED ||
+	    handle_to->state == AMOD_STATE_UNDEFINED) {
+		LOG_DBG("A module is in an invalid state for connecting modules %s to %s",
+			handle_from->name, handle_to->name);
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&handle_from->dest_mutex, K_FOREVER);
+
+	if (handle_from == handle_to) {
+		handle_from->data_tx = true;
+		handle_from->dest_count += 1;
+		LOG_DBG("Return the output of %s on it's TX message queue", handle_from->name);
+	} else {
+		SYS_SLIST_FOR_EACH_CONTAINER(&handle_from->hdl_dest_list, handle, node) {
+			if (handle_to == handle) {
+				LOG_DBG("Already attached %s to %s", handle_to->name,
+					handle_from->name);
+				return -EALREADY;
+			}
+		}
+
+		sys_slist_append(&handle_from->hdl_dest_list, &handle_to->node);
+		handle_from->dest_count += 1;
+
+		LOG_DBG("Connected the output of %s to the input of %s", handle_from->name,
+			handle_to->name);
+	}
+
+	k_mutex_unlock(&handle_from->dest_mutex);
+
+	return 0;
+}
+
+/**
+ * @brief Function to disconnect modules from each other.
+ *
+ */
+int audio_module_disconnect(struct audio_module_handle *handle,
+			    struct audio_module_handle *handle_disconnect)
+{
+	if (handle == NULL || handle_disconnect == NULL) {
+		LOG_DBG("Module handle is NULL");
+		return -EINVAL;
+	}
+
+	if (handle->description->type == AMOD_TYPE_UNDEFINED ||
+	    handle->description->type == AMOD_TYPE_OUTPUT ||
+	    handle_disconnect->description->type == AMOD_TYPE_UNDEFINED ||
+	    handle_disconnect->description->type == AMOD_TYPE_INPUT) {
+		LOG_DBG("Disonnection of modules %s from %s, is not supported",
+			handle_disconnect->name, handle->name);
+		return -ENOTSUP;
+	}
+
+	if (handle->state == AMOD_STATE_UNDEFINED ||
+	    handle_disconnect->state == AMOD_STATE_UNDEFINED) {
+		LOG_DBG("A module is an invalid state or type for disconnection");
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&handle->dest_mutex, K_FOREVER);
+
+	if (handle == handle_disconnect) {
```

Comment:
Can these be the same? And, I see two `handle->dest_count -= 1;` here, which is odd to me.

Reply:
Yes as this indicates that the module is putting audio data on it's TX queue. Moved the decrement so as only one.